src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -470,28 +471,34 @@ public String canonicalize() {
      */
     private String canonicalize(boolean coordinatesOnly) {
         final StringBuilder purl = new StringBuilder();
-        purl.append(SCHEME_PART).append(type).append("/");
+        purl.append(SCHEME_PART).append(type).append('/');
         if (namespace != null) {
             purl.append(encodePath(namespace));
-            purl.append("/");
+            purl.append('/');
         }
         purl.append(percentEncode(name));
         if (version != null) {
-            purl.append("@").append(percentEncode(version));
+            purl.append('@').append(percentEncode(version));
         }
-        if (! coordinatesOnly) {
+
+        if (!coordinatesOnly) {
             if (qualifiers != null) {
-                purl.append("?");
-                qualifiers.forEach((key, value) -> {
-                    purl.append(toLowerCase(key));
-                    purl.append("=");
-                    purl.append(percentEncode(value));
-                    purl.append("&");
-                });
-                purl.setLength(purl.length() - 1);
+                purl.append('?');
+                Iterator<Map.Entry<String, String>> it = qualifiers.entrySet().iterator();
+                Map.Entry<String, String> entry = it.next();
+                purl.append(entry.getKey());
+                purl.append('=');
+                purl.append(percentEncode(entry.getValue()));
+                while (it.hasNext()) {
+                    purl.append('&');
+                    entry = it.next();
+                    purl.append(toLowerCase(entry.getKey()));
+                    purl.append('=');
+                    purl.append(percentEncode(entry.getValue()));
+                }
```

Comment:
Why this big change?
`Map.forEach` is garbage-free, while using an iterator isn't.

Reply:
I am not sure which part is the "garbage". It seemed like a bit of a hack to write an extra `'&'` and then delete it by filling the rest of the buffer with `'\0'`. I thought this was better as it just writes the exact number. We know we must have at least one pair at at that point.

I feel like something like

```java
qualifiers.stream().entrySet().stream().map(Object::toString).collect(Collectors.joining("&"));
```

looks even "cleaner", but requires multiple buffers. I guess the trick is to do it using a single buffer.

But, I could revert it. I don't care that much.

Reply:
By "garbage-free" I mean that it does not create any temporary Java objects (except those created by `toLowerCase` and `percentEncode`).

The new code creates an iterator and map entries.

Reply:
I see, you meant "garbage" as in "garbage collection", but `Map::forEach` _does_ create an `entrySet` already. Point taken about the iterator, though. It's possible to remove the iterator.

Reply:
`Map.forEach` is not GC-free but some implementations (e.g. `HashMap`) are.

Reply:
Yes, it seems there is no `entrySet` in `TreeMap` until you create it. In that case, `validateQualifiers()` should remove its use too?

Reply:
My rule of thumb is that if it is not too much trouble, GC-free code should be preferred. If it makes code much more complex, then more modern constructs are preferred.

As a curiosity IDEA always suggests using `forEach()` instead of `stream().forEach()`

Reply:
@ppkarwasz Since the `entrySet` may have been created already in a method prior to the `toString()` call, I think it exists for the lifetime of the object, and there should be no additional overhead (memory or time) to using it in `toString()`.